tests/integration/security/mtls_cert_securenaming/mtls_cert_securenaming_test.go

```diff
@@ -0,0 +1,244 @@
+				Prefix: "test-mtls",
+				Inject: true,
+			})
+			namespace.ClaimOrFail(t, ctx, istioCfg.SystemNamespace)
+			// Check that the CA certificate in the configmap of each namespace is as expected, which
+			// is used for data plane to control plane TLS authentication.
+			retry.UntilSuccessOrFail(t, func() error {
+				return checkCACert(ctx, t, testNamespace)
+			}, retry.Delay(time.Second), retry.Timeout(10*time.Second))
+			var a, b echo.Instance
+			echoboot.NewBuilderOrFail(ctx, ctx).
+				With(&a, util.EchoConfig("a", testNamespace, false, nil, g, p)).
+				With(&b, util.EchoConfig("b", testNamespace, false, nil, g, p)).
+				BuildOrFail(t)
+
+			ctx.NewSubTest("mTLS cert validation").
+				Run(func(ctx framework.TestContext) {
+					// Verify that the certificate issued to the sidecar is as expected.
+					connectTarget := fmt.Sprintf("b.%s:80", testNamespace.Name())
+					out, err := cert.DumpCertFromSidecar(testNamespace, "app=a", "istio-proxy",
+						connectTarget)
+					if err != nil {
+						t.Fatalf("failed to dump certificate: %v", err)
+						return
+					}
+					verifyCertificates(t, out)
+
+					// Verify mTLS works between a and b
+					callOptions := echo.CallOptions{
+						Target:   b,
+						PortName: "http",
+						Scheme:   scheme.HTTP,
+					}
+					checker := connection.Checker{
+						From:          a,
+						Options:       callOptions,
+						ExpectSuccess: true,
+					}
+					checker.CheckOrFail(ctx)
+				})
+
+			secureNamingTestCases := []struct {
+				name            string
+				destinationRule string
+				expectSuccess   bool
+			}{
+				{
+					name:            "connection fails when DR doesn't match SA",
+					destinationRule: defaultIdentityDR,
+					expectSuccess:   false,
+				},
+				{
+					name:            "connection succeeds when DR matches SA",
+					destinationRule: correctIdentityDR,
+					expectSuccess:   true,
+				},
+				{
+					name:            "connection fails when DR contains non-matching, non-existing SA",
+					destinationRule: nonExistIdentityDR,
+					expectSuccess:   false,
+				},
+				{
+					name:            "connection succeeds when SA is in the list of SANs",
+					destinationRule: identityListDR,
+					expectSuccess:   true,
+				},
+			}
+			for _, tc := range secureNamingTestCases {
+				ctx.NewSubTest(tc.name).
+					Run(func(ctx framework.TestContext) {
+						dr := strings.ReplaceAll(tc.destinationRule, "NS", testNamespace.Name())
+						g.ApplyConfigOrFail(t, testNamespace, dr)
+						// Verify mTLS works between a and b
+						callOptions := echo.CallOptions{
+							Target:   b,
+							PortName: "http",
+							Scheme:   scheme.HTTP,
+						}
+						checker := connection.Checker{
+							From:          a,
+							Options:       callOptions,
+							ExpectSuccess: tc.expectSuccess,
+						}
+						if err := retry.UntilSuccess(
+							checker.Check, retry.Delay(time.Second), retry.Timeout(5*time.Second)); err != nil {
```

Comment:
I would make the timeout higher than 5s to account for weird case where the config distribution is slow, and add `retry.Converge(5)` or similar to avoid false positives. We may be calling existing connections so the tls changes don't apply?

Reply:
Sure. My thought is that when CDS changes, the corresponding connections are reestablished. That's from my observation through the tests, didn't look at the Envoy code, though.